File: QViz.Java/src/test/java/com/periscope/qviz/test/APIClientTests.java
package com.periscope.qviz.test;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.periscope.qviz.client.APIClient;
import com.periscope.qviz.json.*;
import org.apache.http.util.EntityUtils;
import org.junit.jupiter.api.*;

import java.text.SimpleDateFormat;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public class APIClientTests {

	private static String apiURL = "https://api-demo.qviz.io";
	private static AccessUser apiUser;
	private static ObjectMapper apiMapper;
	private static String apiResponse;
	private static String companyId = "05a0ab39-5198-4ae4-aa29-0ae346db03e2";
	private static User user;

	@BeforeAll
	public static void getAccessToken() throws Exception {
		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS");
		apiMapper = new ObjectMapper();
		apiMapper.setDateFormat(dateFormat);
		Authentication authentication = new Authentication("admin", "password");
		APIClient apiClient = new APIClient(apiURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.body = apiMapper.writeValueAsString(authentication);
		if (apiClient.post("/api/authentication") == 200) {
			apiResponse = EntityUtils.toString(apiClient.response.getEntity());
			apiUser = apiMapper.readValue(apiResponse, AccessUser.class);
		} else {
			throw apiClient.getError();
		}
	}

	@Test
	@Order(1)
	@DisplayName("Get all the Users from QViz")
	public void getAllUsers() throws Exception {
		APIClient apiClient = new APIClient(apiURL);
		apiClient.queries.add(new KeyValue("companyId", companyId));
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + apiUser.access_token));
		if (apiClient.get("/api/Users") == 200) {
			apiResponse = EntityUtils.toString(apiClient.response.getEntity());
			List<User> allUsers = apiMapper.readValue(apiResponse, new TypeReference<List<User>>() {
			});
			assertTrue(allUsers.size() > 0, "There should at least be one User");
		} else {
			fail("API Status Code (" +
				apiClient.response.getStatusLine().getStatusCode() +
				") does not match 200.\nAPI Error Message: " +
				apiClient.getError().getMessage()
			);
		}
	}

	@Test
	@Order(2)
	@DisplayName("Add a new User to QViz")
	public void addUser() throws Exception {
		user = new User();
		user.userId = UUID.randomUUID().toString();
		user.companyId = companyId;
		user.username = "Sri";
		user.password = "Sri@2019";
		user.firstName = "Srinivasan";
		user.lastName = "Annamalai";
		user.email = "email@sri.ink";
		APIClient apiClient = new APIClient(apiURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + apiUser.access_token));
		apiClient.headers.add(new KeyValue("userId", apiUser.userId));
		apiClient.body = apiMapper.writeValueAsString(user);
		if (apiClient.post("/api/Users") == 200) {
			apiResponse = EntityUtils.toString(apiClient.response.getEntity());
			UserResponse response = apiMapper.readValue(apiResponse, UserResponse.class);
			user = response.value;
			user.username = user.name;
			assertEquals(
				"User created successfully",
				response.message,
				"User addition should be successful"
			);
		} else {
			fail("API Status Code (" +
				apiClient.response.getStatusLine().getStatusCode() +
				") does not match 200.\nAPI Error Message: " +
				apiClient.getError().getMessage()
			);
		}
	}

	@Test
	@Order(3)
	@DisplayName("Get an existing User from QViz")
	public void getUser() throws Exception {
		APIClient apiClient = new APIClient(apiURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + apiUser.access_token));
		if (apiClient.get("/api/Users/" + user.userId) == 200) {
			apiResponse = EntityUtils.toString(apiClient.response.getEntity());
			User oldUser = apiMapper.readValue(apiResponse, User.class);
			assertEquals(
				user.userId,
				oldUser.userId,
				"Fetched User ID should match the one provided"
			);
		} else {
			fail("API Status Code (" +
				apiClient.response.getStatusLine().getStatusCode() +
				") does not match 200.\nAPI Error Message: " +
				apiClient.getError().getMessage()
			);
		}
	}

	@Test
	@Order(4)
	@DisplayName("Update an existing User from QViz")
	public void updateUser() throws Exception {
		APIClient apiClient = new APIClient(apiURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + apiUser.access_token));
		apiClient.headers.add(new KeyValue("userId", apiUser.userId));
		user.email = "email@srinivasan.pro";
		apiClient.body = apiMapper.writeValueAsString(user);
		if (apiClient.put("/api/Users/" + user.userId) == 200) {
			apiClient.get("/api/Users/" + user.userId); //Put not returning the user object hence calling the get
			apiResponse = EntityUtils.toString(apiClient.response.getEntity());
			User updatedUser = apiMapper.readValue(apiResponse, User.class);
			assertEquals(
				user.email,
				updatedUser.email,
				"Updated User Email should match the one provided"
			);
		} else {
			fail("API Status Code (" +
				apiClient.response.getStatusLine().getStatusCode() +
				") does not match 200.\nAPI Error Message: " +
				apiClient.getError().getMessage()
			);
		}
	}

	@Test
	@Order(5)
	@DisplayName("Delete an existing User from QViz")
	public void deleteUser() throws Exception {
		APIClient apiClient = new APIClient(apiURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + apiUser.access_token));
		if (apiClient.delete("/api/Users/" + user.userId) == 200) {
			apiResponse = EntityUtils.toString(apiClient.response.getEntity());
			UserResponse response = apiMapper.readValue(apiResponse, UserResponse.class);
			assertEquals(
				"User deleted successfully",
				response.message,
				"User deletion should be successful"
			);
		} else {
			fail("API Status Code (" +
				apiClient.response.getStatusLine().getStatusCode() +
				") does not match 200.\nAPI Error Message: " +
				apiClient.getError().getMessage()
			);
		}
	}

}

File: QViz.NET/Periscope.QViz.Test/APIClientTests.cs
using System;
using System.Collections.Generic;
using NUnit.Framework;
using Periscope.QViz.JSON;
using Periscope.QViz.Clients;

namespace Periscope.QViz.Test
{
	[TestFixture]
	public class APIClientTests
	{
		private const string APIURL = "https://api-demo.qviz.io";
		private const string CompanyId = "05a0ab39-5198-4ae4-aa29-0ae346db03e2";
		private static AccessUser _apiUser;
		private static User _user;

		[OneTimeSetUp]
		public static void GetAccessToken()
		{
			var authentication = new Authentication("admin", "password");
			var apiClient = new APIClient(APIURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Body = authentication;
			_apiUser = apiClient.Post<AccessUser>("/api/authentication");
			if (!apiClient.Response.IsSuccessful)
			{
				throw apiClient.GetError();
			}
		}

		[Test(Description = "Get all the Users from QViz")]
		[Order(1)]
		public void GetAllUsers()
		{
			var apiClient = new APIClient(APIURL);
			apiClient.Queries.Add(new KeyValue("companyId", CompanyId));
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _apiUser.access_token));
			var allUsers = apiClient.Get<List<User>>("/api/Users");
			if (apiClient.Response.IsSuccessful)
			{
				Assert.True(allUsers.Count > 0, "There should at least be one User");
			}
			else
			{
				Assert.Fail("API Status Code (" + apiClient.Response.StatusCode + ") does not match 200." +
							"\nAPI Error Message: " + apiClient.GetError().Message);
			}
		}

		[Test(Description = "Add a new User to QViz")]
		[Order(2)]
		public void AddUser()
		{
			_user = new User
			{
				userId = Guid.NewGuid().ToString(),
				companyId = CompanyId,
				userName = "Annamalai4",
				name = "Annamalai4",
				password = "Sri@2019",
				firstName = "Srinivasan",
				lastName = "Annamalai",
				email = "srinivasan.annamalai@qviz.io"
			};
			var apiClient = new APIClient(APIURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _apiUser.access_token));
			apiClient.Headers.Add(new KeyValue("userId", _apiUser.userId));
			apiClient.Body = _user;
			var userResponse = apiClient.Post<UserResponse>("/api/Users");
			if (apiClient.Response.IsSuccessful)
			{
				_user = userResponse.value;
				Assert.AreEqual(
					"User created successfully",
					userResponse.message,
					"User addition should be successful"
				);
			}
			else
			{
				Assert.Fail("API Status Code (" + apiClient.Response.StatusCode + ") does not match 200." +
							"\nAPI Error Message: " + apiClient.GetError().Message);
			}
		}

		[Test(Description = "Get an existing User from QViz")]
		[Order(3)]
		public void GetUser()
		{
			var apiClient = new APIClient(APIURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _apiUser.access_token));
			var oldUser = apiClient.Get<User>("/api/Users/" + _user.userId);
			if (apiClient.Response.IsSuccessful)
			{
				Assert.AreEqual(
					_user.userId,
					oldUser.userId,
					"Fetched User ID should match the one provided"
				);
			}
			else
			{
				Assert.Fail("API Status Code (" + apiClient.Response.StatusCode + ") does not match 200." +
							"\nAPI Error Message: " + apiClient.GetError().Message);
			}
		}

		[Test(Description = "Update an existing User from QViz")]
		[Order(4)]
		public void UpdateUser()
		{
			var apiClient = new APIClient(APIURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _apiUser.access_token));
			apiClient.Headers.Add(new KeyValue("userId", _apiUser.userId));
			_user.email = "srinivasan.annamalai@qviz.io";
			_user.userName = string.IsNullOrEmpty(_user.name) ? "Sri" : _user.name;
			_user.name = string.IsNullOrEmpty(_user.userName) ? "Sri" : _user.userName;
			apiClient.Body = _user;
			QVizResponseObject<User> qVizResponse = apiClient.Put<QVizResponseObject<User>>("/api/Users/" + _user.userId);

			if (apiClient.Response.IsSuccessful)
			{
				Assert.AreEqual(
					_user.email,
					qVizResponse.Value.email,
					"Updated User Email should match the one provided"
				);
			}
			else
			{
				Assert.Fail("API Status Code (" + apiClient.Response.StatusCode + ") does not match 200." +
							"\nAPI Error Message: " + apiClient.GetError().Message);
			}
		}

		[Test(Description = "Delete an existing User from QViz")]
		[Order(5)]
		public void DeleteUser()
		{
			var apiClient = new APIClient(APIURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _apiUser.access_token));
			var userResponse = apiClient.Delete<UserResponse>("/api/Users/" + _user.userId);
			if (apiClient.Response.IsSuccessful)
			{
				Assert.AreEqual(
					"User deleted successfully",
					userResponse.message,
					"User deletion should be successful"
				);
			}
			else
			{
				Assert.Fail("API Status Code (" + apiClient.Response.StatusCode + ") does not match 200." +
							"\nAPI Error Message: " + apiClient.GetError().Message);
			}
		}
	}
}